src/components/MagicCodeInput.js

```diff
@@ -0,0 +1,282 @@
+
+    componentWillUnmount() {
+        if (!this.focusTimeout) {
+            return;
+        }
+        clearTimeout(this.focusTimeout);
+    }
+
+    /**
+     * Focuses on the input when it is pressed.
+     *
+     * @param {object} event The event passed by the input.
+     * @param {number} index The index of the input.
+     */
+    onFocus(event, index) {
+        event.preventDefault();
+        this.setState({
+            input: '',
+            focusedIndex: index,
+            editIndex: index,
+        });
+        this.inputRef.focus();
+    }
+
+    /**
+     * Updates the magic inputs with the contents written in the
+     * input. It spreads each number into each input and updates
+     * the focused input on the next empty one, if exists.
+     * It handles both fast typing and only one digit at a time
+     * in a specific position.
+     *
+     * @param {string} value The contents of the input text.
+     */
+    onChangeText(value) {
+        if (_.isUndefined(value) || _.isEmpty(value) || !ValidationUtils.isNumeric(value)) {
+            return;
+        }
+
+        const numbersArr = value.trim().split('');
+        this.setState((prevState) => {
+            const numbers = [
+                ...prevState.numbers.slice(0, prevState.editIndex),
+                ...numbersArr.slice(0, CONST.MAGIC_CODE_NUMBERS - prevState.editIndex),
+            ];
+
+            // Updates the focused input taking into consideration the last input
+            // edited and the number of digits added by the user.
+            const focusedIndex = Math.min(prevState.editIndex + (numbersArr.length - 1), CONST.MAGIC_CODE_NUMBERS - 1);
+
+            return {
+                numbers,
+                focusedIndex,
+                input: value,
+                editIndex: prevState.editIndex,
+            };
+        }, () => {
+            const finalInput = this.composeToString(this.state.numbers);
+            this.props.onChange(finalInput);
+
+            if (this.props.submitOnComplete && finalInput.length === CONST.MAGIC_CODE_NUMBERS) {
+                this.props.onSubmit(finalInput);
+            }
+        });
+    }
+
+    /**
+     * Handles logic related to certain key presses.
+     *
+     * NOTE: when using Android Emulator, this can only be tested using
+     * hardware keyboard inputs.
+     *
+     * @param {object} event The event passed by the key press.
+     */
+    onKeyPress({nativeEvent: {key: keyValue}}) {
+        // Handles the delete character logic if the current input is less than 2 characters,
+        // meaning that it's the last character to be deleted or it's a character being
+        // deleted in the middle of the input, which should delete all the characters after it.
+        if (keyValue === 'Backspace' && this.state.input.length < 2) {
+            this.setState((prevState) => {
+                const numbers = [...prevState.numbers];
+                numbers[prevState.focusedIndex] = '';
```

Comment:
Is the number with the index "focusedIndex" removed in line 190?

Reply:
No, that slice is unnecessary, will remove